src/Kernel/CacheInvalidatingContainer.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Core\Kernel;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+final class CacheInvalidatingContainer implements ContainerInterface {
+    public function __construct(
+        private readonly ContainerInterface $wrapped
+    )
+    {}
+
+    public function set(string $id, ?object $service)
+    {
+        $this->wrapped->set($id, $service);
+    }
+
+    public function get(string $id, int $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE): ?object
+    {
+        try {
+            return $this->wrapped->get($id, $invalidBehavior);
+        } catch (\Throwable $e) {
+            RectorKernel::clearCache();
+
+            throw $e;
```

Comment:
I though about automatically creating a fresh uncached container in this case, but it did not yet work for me.